foursight_core/app_utils.py

```diff
@@ -282,10 +282,11 @@ def get_portal_url(self, env_name: str) -> Optional[str]:
                 portal_url = environment_and_bucket_info.get("fourfront")
                 self._cached_portal_url[env_name] = portal_url
             except Exception as e:
+                if raise_exception:
+                    raise e
                 message = f"Error getting portal URL: {get_error_message(e)}"
                 logger.error(message)
                 return None
```

Comment:
So if there's a `None` we don't cache it. Is that good? It means every time you'll be asking. Is it likely to change? Not sure the use case here, and I get not wanting to cache failure in some contexts, but caches also end up saving you from a lot of wasted net traffic.

Also, somewhat unrelated, does python not have some built-in caching stuff, or do we not? We shouldn't be doing this kind of logic on a one-off basis per function. It should be a decoration operation.

Also, a decorator can have more options that are easy to turn on and off. Like it should be possible to tell the decorator that you want to cache only for a short while so you aren't trying many times a second but perhaps are checking every minute or 15 minutes or whatever reasonable interval you think might make a difference to a human editing it and wanting to see the effect. Nanosecond response time is not needed.

Reply:
Yeah my thought was if we have None, then something is pretty wrong (e.g. expired Portal access key or SSL certificate), and being slow in such an unusual case isn't necessarily very bad. When I first did this I didn't know about the Python @lru_cache caching decorator (I do use it elsewhere in later code), but this does not have the feature of not caching if the return value is None, and running into issues trying to implement that, may keep trying.

Reply:
Also, @lru_cache does not work with structured (dict/list) arguments, which prevents it from being used (in addition to not having ability to not cache None, and not ttl support). Your comments actually made me consider a more capable caching decorator, which I did experiment with and have a draft of in dcictutils PR-251 and, if/when approved, I can use to simplify some of this.